pkg/mcp/internal/queue.go

```diff
@@ -0,0 +1,175 @@
+
+package internal
+
+import (
+	"sync"
+	"time"
+
+	"istio.io/istio/pkg/mcp/env"
+)
+
+// TODO - this can eventually be moved under pkg/mcp/source once the new stack is
+// introduced. The code is temporarily added under pkg/mcp/internal so
+// we can share across package boundaries without exposing outside of the pkg/mcp
+// parent directory.
+
+// UniqueQueue is a specialized queue structure. It has the following properties:
+//
+// - Enqueuing an item already in the queue is idempotent. The first time an item
+//   is added to the queue determines its order in the queue until it is dequeued.
+//   Callers can safely update the queued items state while preserving it's place
+//   in the queue.
+//
+// - Enqueuing an item in the queue creates a schedule event. The caller can select
+//   over the Ready() channel to process this event and remove items from the queue.
+//
+// - The maximum queue depth is fixed.
+//
+// - The queue can be safely closed. The caller is responsible for checking the Done()
+//   state of the queue and should stop checking Ready() and invoking Dequeue() when
+//   the channel returned by Done() is closed.
+//
+// This is intended to be used by the MCP source/server packages for managing
+// per-type watch state.
+type UniqueQueue struct {
+	sync.Mutex
+	doneChan       chan struct{}
+	doneChanClosed bool
+	queuedSet      map[interface{}]struct{}
+	// Enqueue at the tail, Dequeue from the head
+	// head == tail   => empty
+	// head == tail+1 => full
+	head     int
+	tail     int
+	queue    []interface{}
+	maxDepth int
+
+	scheduleChan chan struct{}
+}
+
+// NewUniqueScheduledQueue creates a new unique queue specialized for MCP source/server implementations.
+func NewUniqueScheduledQueue(maxDepth int) *UniqueQueue {
+	return &UniqueQueue{
+		queue:        make([]interface{}, maxDepth+1),
+		queuedSet:    make(map[interface{}]struct{}, maxDepth),
+		maxDepth:     maxDepth,
+		scheduleChan: make(chan struct{}, maxDepth),
+		doneChan:     make(chan struct{}),
+	}
+}
+
+// return the next index accounting for wrap
+func (q *UniqueQueue) inc(idx int) int {
+	idx = idx + 1
+	if idx == q.maxDepth+1 {
+		idx = 0
+	}
+	return idx
+}
+
+// Empty returns true if the queue is empty
+func (q *UniqueQueue) Empty() bool {
+	q.Lock()
+	defer q.Unlock()
+	return q.head == q.tail
+}
+
+// internal version of full() that can be called
+// with the lock held.
+func (q *UniqueQueue) full() bool {
+	return q.head == q.inc(q.tail)
+}
+
+// Full returns true if the queue is full
+func (q *UniqueQueue) Full() bool {
+	q.Lock()
+	defer q.Unlock()
+	return q.full()
+}
+
+// Enqueue an item in the queue. The same item may be safely enqueued multiple
+// times. Attempts to enqueue an already queued item have no affect on the order
+// of already queued items.
+//
+// Returns true if the item exists in the queue upon return. Otherwise,
+// returns false if the item could not be queued.
+func (q *UniqueQueue) Enqueue(w interface{}) bool {
```

Comment:
There doesn't seem to be any protection against caller mischief. Caller can enqueue things that are arbitrary. Especially given that both the key and value are the same.

Would it make sense to have this queue key things based on collection names?